src/concatenate.jl

```diff
@@ -0,0 +1,163 @@
+
+# For now, simply couple back to base implementation
+function Base.axes(concat::Concatenated)
+  catdims = Base.dims2cat(dims(concat))
+  return Base.cat_size_shape(catdims, concat.args...)
+end
+
+# Main logic
+# ----------
+"""
+    Concatenate.cat(args...; dims)
+
+Concatenate the supplied `args` along dimensions `dims`.
+
+See also [`cat!`](@ref).
+"""
+cat(args...; dims) = Base.materialize(concatenated(args...; dims))
+Base.materialize(concat::Concatenated) = copy(concat)
+
+"""
+    Concatenate.cat!(dest, args...; dims)
+
+Concatenate the supplied `args` along dimensions `dims`, placing the result into `dest`.
+"""
+function cat!(dest, args...; dims)
+  Base.materialize!(dest, concatenated(dims, args...))
+  return dest
+end
+Base.materialize!(dest, concat::Concatenated) = copyto!(dest, concat)
+
+Base.copy(concat::Concatenated) = copyto!(similar(concat), concat)
+
+# default falls back to replacing interface with Nothing
+# this permits specializing on typeof(dest) without ambiguities
+# Note: this needs to be defined for AbstractArray specifically to avoid ambiguities with Base.
+@inline Base.copyto!(dest::AbstractArray, concat::Concatenated) =
+  copyto!(dest, convert(Concatenated{Nothing}, concat))
+
+# couple back to Base implementation if no specialization exists:
+# https://github.com/JuliaLang/julia/blob/29da86bb983066dd076439c2c7bc5e28dbd611bb/base/abstractarray.jl#L1852
+function Base.copyto!(dest::AbstractArray, concat::Concatenated{Nothing})
+  catdims = Base.dims2cat(dims(concat))
+  shape = Base.cat_size_shape(catdims, concat.args...)
+  count(!iszero, catdims)::Int > 1 && zero!(dest)
+  return Base.__cat(dest, shape, catdims, concat.args...)
+end
+
+# Array implementation
+# --------------------
+# Write in terms of a generic cat_offset!, which in term aims to specialize on 1 argument
+# at a time via cat_offset1! to avoid having to write too many specializations
+# function cat_offset!(dest, shape, catdims, offsets, x, X...)
+#   dest, newoffsets = cat_offset1!(dest, shape, catdims, offsets, x)
+#   return cat_offset!(dest, shape, catdims, newoffsets, X...)
+# end
+# cat_offset!(dest, shape, catdims, offsets) = dest
+
+# this is the typical specialization point, which is no longer vararg.
+# it simply computes indices and calls out to copy_or_fill!, so if that
+# pattern works you can also overload that function
+# function cat_offset1!(dest, shape, catdims, offsets, x)
+#   inds = ntuple(length(offsets)) do i
+#     (i ≤ length(catdims) && catdims[i]) ? offsets[i] .+ axes(x, i) : 1:shape[i]
+#   end
+#   copy_or_fill!(dest, inds, x)
+#   newoffsets = ntuple(length(offsets)) do i
+#     (i ≤ length(catdims) && catdims[i]) ? offsets[i] + size(x, i) : offsets[i]
+#   end
+#   return dest, newoffsets
+# end
+
+# copy of Base._copy_or_fill!
+# copy_or_fill!(A, inds, x) = fill!(view(A, inds...), x)
+# copy_or_fill!(A, inds, x::AbstractArray) = (A[inds...] = x)
+
+zero!(x::AbstractArray) = fill!(x, zero(eltype(x)))
```

Comment:
It seems like we are going to end up with a few `zero!` definitions in various places. Should we define `DerivableInterfaces.zero!` which is used here and in SparseArraysBase.jl?

As we've discussed, SparseArraysBase.jl currently uses `ArrayLayouts.zero!`, but maybe it would be better to have `DerivableInterfaces.zero!` which we use in our own code, and then types that want to use ArrayLayouts.jl can overload `ArrayLayouts.zero!` (which could just forward to `DerivableInterfaces.zero!`).

Reply:
I definitely think that's a good idea, it's strange to me that it doesn't exist in Base anyways.